<a id='top'></a>

# CSCI3022 S21
# Homework 1: Measures of Centrality and Dispersion
***

**Name**: Clayton Schneider

***

This assignment is due on Canvas by **MIDNIGHT on Monday, February 1**. Your solutions to theoretical questions should be done in Markdown directly below the associated question.  Your solutions to computational questions should include any specified Python code and results as well as written commentary on your conclusions.  Remember that you are encouraged to discuss the problems with your classmates, but **you must write all code and solutions on your own**.

**NOTES**: 

- Any relevant data sets should be available on Canvas. To make life easier on the graders if they need to run your code, do not change the relative path names here. Instead, move the files around on your computer.
- If you're not familiar with typesetting math directly into Markdown then by all means, do your work on paper first and then typeset it later.  Here is a [reference guide](https://math.meta.stackexchange.com/questions/5020/mathjax-basic-tutorial-and-quick-reference) linked on Canvas on writing math in Markdown. **All** of your written commentary, justifications and mathematical work should be in Markdown.  I also recommend the [wikibook](https://en.wikibooks.org/wiki/LaTeX) for LaTex.
- Because you can technically evaluate notebook cells is a non-linear order, it's a good idea to do **Kernel $\rightarrow$ Restart & Run All** as a check before submitting your solutions.  That way if we need to run your code you will know that it will work as expected. 
- It is **bad form** to make your reader interpret numerical output from your code.  If a question asks you to compute some value from the data you should show your code output **AND** write a summary of the results in Markdown directly below your code. 
- 95 points of this assignment are in problems.  The remaining 5 are for neatness, style, and overall exposition of both code and text.
- This probably goes without saying, but... For any question that asks you to calculate something, you **must show all work and justify your answers to receive credit**. Sparse or nonexistent work will receive sparse or nonexistent credit. 

---
**Shortcuts:**  [Problem 1](#p1) | [Problem 2](#p2) | [Problem 3](#p3) |
---

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pylab as plt
%matplotlib inline

[Back to top](#top)

<br>

<a/ id='p1'></a>

### (15 points) Problem 1: Theory and Computation (Means and Medians)
***

A method to investigate the sensitivity of the sample mean and sample median to extreme outliers and changes in the dataset is to replace one or more elements in a given dataset by a number $y$ and investigate the effect when $y$ changes. To illustrate this, consider the following dataset:
 
$$  12.2 \quad -5.0 \quad 1.0 \quad 3.8 \quad -4.1 \quad 5.9 \quad 1.9 \quad 9.0 \quad y$$




**Part A:** Compute the sample mean and sample median for $y=3.5$. Then compute both quantities again for $y=6$.

In [2]:
data1 = [12.2, -5.0, 1.0, 3.8, -4.1, 5.9, 1.9, 9.0, 3.5]

mean1 = np.mean(data1)
print("Mean 1:", mean1)

data2 = [12.2, -5.0, 1.0, 3.8, -4.1, 5.9, 1.9, 9.0, 6]

mean2 = np.mean(data2)
print("Mean 2:", mean2)

Mean 1: 3.1333333333333337
Mean 2: 3.4111111111111114


**Part B:** Is there a value for $y$ that would make the mean of the data equal to 6? If so, calculate the value of $y$ that makes the mean equal to 6. If not, clearly explain why not.
    
Is there a value for $y$ that would make the median of the data equal to 6? If so, calculate the values of $y$ that makes the median equal to 6. If not, clearly explain why not.

In [3]:
data = [12.2, -5.0, 1.0, 3.8, -4.1, 5.9, 1.9, 9.0]

mean = 6

totalsum = len(data)*mean
datasum = 0
for num in data:
    datasum += num

y = totalsum - datasum

print("Missing value:", y)

Missing value: 23.299999999999997


**Part C**: Compute the sample variance and the sample standard deviation for the original dataset given in part A, with $y=6$.

In [4]:
truedata = [12.2, -5.0, 1.0, 3.8, -4.1, 5.9, 1.9, 9.0, 6]
print("Variance:", np.var(truedata))
print("STD:", np.std(truedata))

Variance: 28.5320987654321
STD: 5.341544604834083


**Part D**: Compute both the sample mean and median for the following cases: 
- $y=9$ 
- $y=50$ 
- $y=4.36$ 
- $y \to \infty$ 
- $y \to -\infty$

In [9]:
truedata = [12.2, -5.0, 1.0, 3.8, -4.1, 5.9, 1.9, 9.0, 9]
print("Mean:", np.mean(truedata))
print("Median:", np.median(truedata))
print('\n')

truedata = [12.2, -5.0, 1.0, 3.8, -4.1, 5.9, 1.9, 9.0, 50]
print("Mean:", np.mean(truedata))
print("Median:", np.median(truedata))
print('\n')

truedata = [12.2, -5.0, 1.0, 3.8, -4.1, 5.9, 1.9, 9.0, 4.36]
print("Mean:", np.mean(truedata))
print("Median:", np.median(truedata))
print('\n')

truedata = [12.2, -5.0, 1.0, 3.8, -4.1, 5.9, 1.9, 9.0, np.inf]
print("Mean:", np.mean(truedata))
print("Median:", np.median(truedata))
print('\n')

truedata = [12.2, -5.0, 1.0, 3.8, -4.1, 5.9, 1.9, 9.0, np.inf*-1]
print("Mean:", np.mean(truedata))
print("Median:", np.median(truedata))
print('\n')

Mean: 3.744444444444445
Median: 3.8


Mean: 8.3
Median: 3.8


Mean: 3.228888888888889
Median: 3.8


Mean: inf
Median: 3.8


Mean: -inf
Median: 1.9




**Part E**: Think about the previous parts, above, and describe in words or mathematical notation the answers to the following two questions:

- By varying $y$, what is the set of all the possible values that the sample mean could take on?
- By varying $y$, what is the set of all the possible values that the sample median could take on? Specifically, for what sets of $y$ values does the median take on its different possible values?

- mean: [-infinity to infinity]
- median: [y, 1.9, 3.8]

**Part F**: Describe in words or mathematical notation, what happens to the sample standard deviation when $y$ is varied in the following ways: 
 
- $y \to \infty$ 
- $y \to \bar{x}$

In [10]:
truedata = [12.2, -5.0, 1.0, 3.8, -4.1, 5.9, 1.9, 9.0,  3.0875000000000004]
print("Mean:", np.mean(truedata))
print("STD:", np.std(truedata))

Mean: 3.0875000000000004
STD: 5.262537938633369


- STD $\rightarrow$ infinity
- STD $\rightarrow$ 5.2625

[Back to top](#top)
<a/ id='p2'></a>

## (15 pts) Problem 2: Theory and Computation (Scaling)
***
Consider the following 3 data sets:

`A=[0,1,2,3,4,5,6,7,8,9,10,11,12]`

`B=[0,0,0,12,7,18,47,25,0,13,0,35]`

`C` is the random data set generated by using `np.random.exponential(scale=43, size=1000)`

For each data set, perform the following computations in parts A, B, and C:

**Part A:** Compute and print the mean and standard deviation of the data set.

In [11]:
A=[0,1,2,3,4,5,6,7,8,9,10,11,12]
B=[0,0,0,12,7,18,47,25,0,13,0,35]
C=np.random.exponential(scale=43, size=1000)

print("Mean of A:", np.mean(A))
print("STD of A:", np.std(A))
print("Mean of B:", np.mean(B))
print("STD of B:", np.std(B))
print("Mean of C:", np.mean(C))
print("STD of C:", np.std(C))

Mean of A: 6.0
STD of A: 3.7416573867739413
Mean of B: 13.083333333333334
STD of B: 14.974747261825675
Mean of C: 43.45465833402732
STD of C: 42.20030484891801


**Part B:** Compute and print  the mean and standard deviation of the new data set formed by subtracting the original mean from each observation.

In [12]:
Anorm = np.subtract(A, np.mean(A))
Bnorm = np.subtract(B, np.mean(B))
Cnorm = np.subtract(C, np.mean(C))

print("Mean of A:", np.mean(Anorm))
print("STD of A:", np.std(Anorm))
print("Mean of B:", np.mean(Bnorm))
print("STD of B:", np.std(Bnorm))
print("Mean of C:", np.mean(Cnorm))
print("STD of C:", np.std(Cnorm))

Mean of A: 0.0
STD of A: 3.7416573867739413
Mean of B: -1.1842378929335002e-15
STD of B: 14.974747261825675
Mean of C: 2.5011104298755525e-15
STD of C: 42.20030484891801


**Part C:** Compute and print  the mean and standard deviation of the new data set formed by subtracting the original mean from each observation and then dividing by the original standard deviation.

In [13]:
Anormscale = np.divide(Anorm, np.std(A))
Bnormscale = np.divide(Bnorm, np.std(B))
Cnormscale = np.divide(Cnorm, np.std(C))

print("Mean of A:", np.mean(Anormscale))
print("STD of A:", np.std(Anormscale))
print("Mean of B:", np.mean(Bnormscale))
print("STD of B:", np.std(Bnormscale))
print("Mean of C:", np.mean(Cnormscale))
print("STD of C:", np.std(Cnormscale))

Mean of A: -5.124106267500723e-17
STD of A: 1.0
Mean of B: -9.251858538542972e-17
STD of B: 1.0
Mean of C: 6.394884621840901e-17
STD of C: 1.0


**Part D:** Why might this result matter?

This is a convenient and easy-to-understand way to normalize our data. Not only have we centered the mean around 0, but we have also scaled it such that the standard deviation is (almost) exactly 1. This can be very important to machine learning applications, or other instances where it may just make sense to plot the data around a baseline average of 0. 

**Part E:** Looking at each of the 3 data sets, come up with a real-world context where those spectrum of numbers might make for reasonable observations.

- A: counting the seconds (from start) where a Bluetooth connection is maintained 
- B: number of students absent in each consecutive class lecture (haha)
- C: accelerometer data on the x-axis from an iPhone tumbling down A-basin

**Part F:** Let's prove a generalization of the result you likely observed in part B.  What happens if you take data set and "recenter" it by adding or subtracting the same number $c$ to each and every observation?

In other words, show that the mean of the data set $Y$ given by $Y_i=X_i+c$ is $\bar{Y}=\bar{X}+c$.

(**Hint**: start with the definition of the mean: $$\bar{Y}=\frac{\sum_{i=1}^n Y_i}{n},$$ then work with the right hand side!)

In this case, the sample mean $\bar{Y}$ would just be moving by $\frac{c}{n}$ on each iteration through $Y$, a total of $n$ times, and so eventually the difference between the old $\bar{Y}$ and the new $\bar{Y}$ would be $c$ once you've parsed the entirety of $Y$. 

[Back to top](#top)
<a/ id='p3'></a>

## (15 pts) Problem 3: Computation (Streaming Means)
***

Data science is often divided into two categories: questions of *what* the best value might be to repreesnt a data problem, and questions of *how* to compute that data value.  Question 1 - and prior lectures - should tell you that computing the mean is valuable!  But *how* do we compute the mean?

Let $x_1, x_2, \ldots, x_n$ be $n$ observations of a variable of interest.  Recall that the sample mean $\bar{x}_n$ and sample variance $s^2_n$ are given by 
<a id='eq1'></a>
$$
\bar{x}_n = \frac{1}{n}\sum_{k=1}^n x_k \quad \textrm{and} \quad s^2_n = \frac{1}{n-1}\sum_{k=1}^n \left( x_k - \bar{x}_n\right)^2 \qquad \tag{Equation 1}
$$

**Part A**:

How many computations - floating point operations: addition, subtraction, multiplication, division each count as 1 operation - are required to compute the mean of the data set with $n$ observations?


2n+1

On each iter:
- add to sum 
- increment n

Once:
- divide by n

**Part B**:

Now suppose our data is *streaming*- we slowly add observations one at a time, instead of seeing the entire data set at once.  We are still interested in the mean, so if we stream the data set `[4,6,0,10, ...]`, we first compute the mean of the the first data point `[4]`, then we recompute the mean of the first two points `[4,6]`, then we recompute the mean of three `[4,6,0]`, and so forth.

Suppose we recompute the mean from scratch after each and every one of our $n$ observations are one-by-one added to our data set.  How many floating point operations are spent computing (and re-computing) the mean of the data set?

$\frac{n(n+1)}{2}$

We should be convinced that streaming a mean costs a lot more computer time than just computing once!

In this problem we explore a smarter method for such an _online_ computation of the mean.  

**Result**: The following relation holds between the mean of the first $n-1$ observations and the mean of all $n$ observations: 

$$
\bar{x}_n = \bar{x}_{n-1} + \frac{x_n - \bar{x}_{n-1}}{n}
$$


A proof of this result is in the [Appendix](#Appendix) after this problem, and requires some careful manipulations of the sum $\bar{x}_n$.  Your task will be to computationally verify and utilize this result.

**Part C**: Write a function `my_sample_mean` that takes as its input a numpy array and returns the mean of that numpy array using the formulas from class ([Equation 1](#eq1)). Write another function `my_sample_var` that takes as its input a numpy array and returns the variance of that numpy array, again using the formulas from class ([Equation 1](#eq1)). You may **not** use any built-in sample mean or variance functions.

In [14]:
def my_sample_mean(arr):
    sum = 0
    
    for item in arr:
        sum += item
        
    mean = sum / len(arr)
    
    return mean

def my_sample_var(arr):
    mean = my_sample_mean(arr)
    
    sum = 0
    
    for item in arr:
        sum += (item - mean)**2
        
    variance = sum / len(arr)-1
    
    return variance

**Part D**: Use your functions from Part C to compute the sample mean and sample variance of the following array, which contains the minutes late that the BuffBus is running on Friday afternoon.

`bus = [312, 4, 10, 0, 22, 39, 81, 19, 8, 60, 80, 42]`

In [15]:
bus = [312, 4, 10, 0, 22, 39, 81, 19, 8, 60, 80, 42]

print("Mean buff-bus-bad:", my_sample_mean(bus))
print("Var buff-bus-bad:", my_sample_var(bus))

Mean buff-bus-bad: 56.416666666666664
Var buff-bus-bad: 6667.409722222222


**Part E**: Implement a third function called `update_mean` that implements the formula discussed after part B. Note that this function will need to take as its input three things: $x_n$, $\bar{x}_{n-1}$ and $n$, and returns $\bar{x}_{n}$. A function header and return statement are provided for you. This function may be auto-graded, so please do not change the given header API - the order of inputs matters! If you change it, you might lose points.

Use this function to compute the values that you get from taking the mean of the first buff buses' lateness, the first two buff buses' lateness, the first three buff buses' lateness, and so on up to all of the `bus` data points from **Part D**. Store your streaming bus means in a numpy array called `buffbus_bad_means`.  Report all 12 estimates in `buffbus_bad_means`.

In [16]:
# Given API:
def update_mean(prev_mean, xn, n):
    now_mean = prev_mean + ((xn - prev_mean)/n)
    return now_mean

buffbus_bad_means = []

prev = 0
for i in range(len(bus)):
    buffbus_bad_means.append(update_mean(prev, bus[i], i+1))
    prev = buffbus_bad_means[-1]
    
print(buffbus_bad_means)

[312.0, 158.0, 108.66666666666666, 81.5, 69.6, 64.5, 66.85714285714286, 60.875, 55.0, 55.5, 57.72727272727273, 56.416666666666664]


<a id='Appendix'></a>

## Appendix 

*Goal*: Prove that 
$$
\bar{x}_n = \bar{x}_{n-1} + \frac{x_n - \bar{x}_{n-1}}{n}
$$

Note that you can get an expression for $\bar{x}_{n-1}$ by simply replacing $n$ in Equation 1 above with $n-1$.

We'll start with $\bar{x}_n$ and massage it until we get the righthand side of the formula

\begin{eqnarray}
\nonumber \bar{x}_n &=& \frac{1}{n} \sum_{k=1}^n x_k \\
&=& \frac{1}{n} \sum_{k=1}^{n-1} x_k + \frac{1}{n}x_n \\
&=& \frac{n-1}{n-1}\frac{1}{n} \sum_{k=1}^{n-1} x_k + \frac{1}{n}x_n \\
&=& \frac{n-1}{n} \left(\frac{1}{n-1} \sum_{k=1}^{n-1} x_k\right) + \frac{1}{n}x_n \\
&=& \frac{n-1}{n} \bar{x}_{n-1} + \frac{1}{n}x_n \\
&=& \frac{n}{n}\bar{x}_{n-1} - \frac{1}{n}\bar{x}_{n-1} + \frac{1}{n}x_n \\
&=&  \bar{x}_{n-1} + \frac{x_n - \bar{x}_{n-1}}{n} \quad \checkmark
\end{eqnarray}

